cmd/operator-sdk/add/api.go

```diff
@@ -20,90 +20,261 @@ import (
+
+  $ operator-sdk add api \
+  --helm-chart=myrepo/app \
+  --helm-chart-version=1.2.3
+
+  $ operator-sdk add api \
+  --helm-chart=app \
+  --helm-chart-repo=https://charts.mycompany.com/
+
+  $ operator-sdk add api \
+  --helm-chart=app \
+  --helm-chart-repo=https://charts.mycompany.com/ \
+  --helm-chart-version=1.2.3
+
+  $ operator-sdk add api \
+  --helm-chart=/path/to/local/chart-directories/app/
+
+  $ operator-sdk add api \
+  --helm-chart=/path/to/local/chart-archives/app-1.2.3.tgz
 `,
 		RunE: apiRun,
 	}
 
-	apiCmd.Flags().StringVar(&apiVersion, "api-version", "",
-		"Kubernetes APIVersion that has a format of $GROUP_NAME/$VERSION (e.g app.example.com/v1alpha1)")
-	if err := apiCmd.MarkFlagRequired("api-version"); err != nil {
-		log.Fatalf("Failed to mark `api-version` flag for `add api` subcommand as required")
-	}
-	apiCmd.Flags().StringVar(&kind, "kind", "", "Kubernetes resource Kind name. (e.g AppService)")
-	if err := apiCmd.MarkFlagRequired("kind"); err != nil {
-		log.Fatalf("Failed to mark `kind` flag for `add api` subcommand as required")
-	}
-	apiCmd.Flags().BoolVar(&skipGeneration, "skip-generation", false,
-		"Skip generation of deepcopy and OpenAPI code and OpenAPI CRD specs")
-	apiCmd.Flags().StringVar(&crdVersion, "crd-version", gencrd.DefaultCRDVersion,
-		"CRD version to generate")
+	// Initialize flagSet struct with command flags
+	apiFlags.AddTo(apiCmd.Flags())
 
 	return apiCmd
 }
 
 func apiRun(cmd *cobra.Command, args []string) error {
+
 	projutil.MustInProjectRoot()
 
-	// Only Go projects can add apis.
-	if err := projutil.CheckGoProjectCmd(cmd); err != nil {
+	operatorType := projutil.GetOperatorType()
+	if operatorType == projutil.OperatorTypeUnknown {
+		return projutil.ErrUnknownOperatorType{}
+	}
+	// Verify the incoming flags.
+	if err := apiFlags.VerifyCommonFlags(operatorType); err != nil {
+		return err
+	}
+
+	log.Infof("Generating api version %s for kind %s.", apiFlags.APIVersion, apiFlags.Kind)
+
+	switch operatorType {
+	case projutil.OperatorTypeGo:
+		if err := doGoAPIScaffold(); err != nil {
+			return err
+		}
+	case projutil.OperatorTypeAnsible:
+		if err := doAnsibleAPIScaffold(); err != nil {
+			return err
+		}
+	case projutil.OperatorTypeHelm:
+		if err := doHelmAPIScaffold(); err != nil {
+			return err
+		}
+	default:
+	}
+	log.Info("API generation complete.")
+	return nil
+}
+
+// TODO
+// Consolidate scaffold func to be used by both "new" and "add api" commands.
+func doAnsibleAPIScaffold() error {
+	// Create and validate new resource.
+	r, err := scaffold.NewResource(apiFlags.APIVersion, apiFlags.Kind)
+	if err != nil {
+		return err
+	}
+	absProjectPath := projutil.MustGetwd()
+	cfg := &input.Config{
+		AbsProjectPath: absProjectPath,
+	}
+	roleFiles := ansible.RolesFiles{Resource: *r}
+	roleTemplates := ansible.RolesTemplates{Resource: *r}
+
+	// update watch.yaml for the given resource r.
+	if err := ansible.UpdateAnsibleWatchForResource(r, absProjectPath); err != nil {
```

Comment:
Ughh, sorry @bharathi-tenneti the rabbit hole continues. This is because [we're doing the role directory expansion during YAML unmarshalling](https://github.com/operator-framework/operator-sdk/blob/29b1d3e2bc9acd2ddfed05ae1a0a01da11bf5611/pkg/ansible/watches/watches.go#L145), which is probably not the right thing to do.

Instead, we might be able to call that function in [this loop](https://github.com/operator-framework/operator-sdk/blob/29b1d3e2bc9acd2ddfed05ae1a0a01da11bf5611/pkg/ansible/watches/watches.go#L291-L303).

Reply:
@joelanford Above fix worked in this scenario. However, broke couple of pre-existing test cases which were written for `pkg/ansible/watches` . They were checking for the unfolded absoulte path for role. Will discuss with @fabianvf whether this will break just the test cases or if the impact is more in some use case.

Reply:
As discussed, with @joelanford and @fabianvf changing the approach back to templating, as using existing Watch struct is posing additional complication while unmarshal.